### ISSUE TITLE
Q: lxc and encrypted rootfs

### ISSUE THREAD

**Issue by huegelc** [lxc/lxc#487]
Hi,

is there a way to create encrypted lxc containers?

**Comment by stgraber** [comment]
Not at this point since we don't support block devices for the container rootfs yet. Once we do (and it's pretty high on our todo), then it should be possible to support luks encrypted block devices.

However I'm not sure I see the point, what are you trying to protect yourself against by using encrypted storage?

The host will always be able to access the content of the container since it would have access to the key and is the one performing the mount. If your fear is about somebody stealing the storage media (HDD or so), then full disk encryption on the media should be faster and won't use any CPU cycles.

**Comment by huegelc** [comment]
We have a root server where we are running a mail server for our customers in a lxc container. At the point of the installation of the rootserver we missed to encrypt the hdd. So we were looking at a option to encrypt the rootfs of the container till we migrate to another server with full hdd encryption. 
I have read this post (https://s3hh.wordpress.com/2012/09/06/ecryptfs-backed-containers/) about encrypting lxc containers but wasn't able to get going with that solution.

**Comment by TommyJerryMairo** [comment]
Hi @stgraber ,

I am also a lxc user who is waiting for encrypted rootfs.
As you might heard, some disks might leak data to the one who pick-up the abandoned disk.
However, encryption can prevent this. If we encrypt the rootfs with LUKS and store keys in TPM or smart cards. Data can never be leaked after we abandon the disk.

In this case, we can run some program which deal with very sensitive informations in lxc

**Comment by hallyn** [comment]
Sorry, what is the issue?  I use encrypted blockdevs for several of my containers.  Now, I think I created them by hand, so I

1. sudo lxc-create <...>

2. luks-setup a new device

3. (these run in a user ns, with separate range each, so I shifted the rootfs's using fuidshift)

4. rsync /usr/lib/lxc/containers/xxx/rootfs into the mounted luksfs device

5. remove /usr/lib/lxc/containers/xxx/rootfs 

6. update the container config to point to the luks device

After each host reboot, I manually hook up the luks device, then start the container.
